### PR TITLE
Sort master VMs by health before upgrade

### PR DESCRIPTION
**Pull request by kargakis**
Ref https://github.com/openshift/openshift-azure/issues/362

@openshift/sig-azure ptal, seems like a quick win until we have the proposal @jim-minter is going to work post-PP.

### CODE REVIEW

**Comment by kargakis** [comment]
I was able to pick up a failed master upgrade with this.

File: pkg/upgrade/update.go [hunk 1]
```diff
@@ -295,9 +301,42 @@ func (u *simpleUpgrader) updateInPlace(ctx context.Context, cs *api.OpenShiftMan
...
+		nodeName := *vm.VirtualMachineScaleSetVMProperties.OsProfile.ComputerName
+		ready, err := masterIsReady(kc, nodeName)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get health for %q: %v", nodeName, err)
+		}
+		if !ready {
+			sorted = append(sorted, vm)
+		} else {
+			if vms[i].Tags == nil {
+				vms[i].Tags = make(map[string]*string)
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/upgrade/update.go]: fixed

File: pkg/upgrade/update.go [hunk 2]
```diff
@@ -224,9 +225,14 @@ func (u *simpleUpgrader) updateInPlace(ctx context.Context, cs *api.OpenShiftMan
 		return err
 	}
 
-	for _, vm := range vms {
+	sorted, err := sortMasterVMsByHealth(vms, cs)
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/upgrade/update.go]: https://github.com/openshift/openshift-azure/issues/382

**Comment by kargakis** [comment]
<details>

```console
INFO[2018-09-10T09:41:52+02:00] convert to internal                           resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] enrich                                        resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] read old config                               resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] merge old and new config                      resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] merging internal data models                  resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] validating internal data models               resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] generating configs                            resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] persist config                                resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] generating arm templates                      resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] write helpers                                 resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] starting update                               resourceGroup=kargakisosa
INFO[2018-09-10T09:41:52+02:00] applying arm template deployment              resourceGroup=kargakisosa
INFO[2018-09-10T09:42:06+02:00] waiting for arm template deployment to complete  resourceGroup=kargakisosa
INFO[2018-09-10T09:42:39+02:00] draining master-000001                        resourceGroup=kargakisosa
INFO[2018-09-10T09:42:39+02:00] drain: node not found, skipping               resourceGroup=kargakisosa
INFO[2018-09-10T09:42:39+02:00] deallocating master-000001 (1)                resourceGroup=kargakisosa
INFO[2018-09-10T09:43:40+02:00] updating master-000001                        resourceGroup=kargakisosa
INFO[2018-09-10T09:44:41+02:00] reimaging master-000001                       resourceGroup=kargakisosa
INFO[2018-09-10T09:45:41+02:00] starting master-000001                        resourceGroup=kargakisosa
INFO[2018-09-10T09:49:49+02:00] waiting for master-000001 to be ready         resourceGroup=kargakisosa
INFO[2018-09-10T09:52:36+02:00] draining master-000000                        resourceGroup=kargakisosa
INFO[2018-09-10T09:52:36+02:00] deallocating master-000000 (0)                resourceGroup=kargakisosa
INFO[2018-09-10T09:55:39+02:00] updating master-000000                        resourceGroup=kargakisosa
INFO[2018-09-10T09:56:40+02:00] reimaging master-000000                       resourceGroup=kargakisosa
INFO[2018-09-10T09:57:41+02:00] starting master-000000                        resourceGroup=kargakisosa
INFO[2018-09-10T10:01:45+02:00] waiting for master-000000 to be ready         resourceGroup=kargakisosa
INFO[2018-09-10T10:03:51+02:00] draining master-000002                        resourceGroup=kargakisosa
INFO[2018-09-10T10:03:51+02:00] deallocating master-000002 (2)                resourceGroup=kargakisosa
INFO[2018-09-10T10:06:54+02:00] updating master-000002                        resourceGroup=kargakisosa
INFO[2018-09-10T10:07:54+02:00] reimaging master-000002                       resourceGroup=kargakisosa
INFO[2018-09-10T10:08:55+02:00] starting master-000002                        resourceGroup=kargakisosa
INFO[2018-09-10T10:13:00+02:00] waiting for master-000002 to be ready         resourceGroup=kargakisosa
INFO[2018-09-10T10:14:15+02:00] setting ss-infra capacity to 3                resourceGroup=kargakisosa
INFO[2018-09-10T10:17:18+02:00] waiting for infra-000005 to be ready          resourceGroup=kargakisosa
INFO[2018-09-10T10:18:26+02:00] draining infra-000002                         resourceGroup=kargakisosa
INFO[2018-09-10T10:19:02+02:00] deleting infra-000002                         resourceGroup=kargakisosa
INFO[2018-09-10T10:21:05+02:00] setting ss-infra capacity to 3                resourceGroup=kargakisosa
INFO[2018-09-10T10:25:08+02:00] waiting for infra-000006 to be ready          resourceGroup=kargakisosa
INFO[2018-09-10T10:26:49+02:00] draining infra-000004                         resourceGroup=kargakisosa
INFO[2018-09-10T10:27:05+02:00] deleting infra-000004                         resourceGroup=kargakisosa
INFO[2018-09-10T10:29:06+02:00] setting ss-compute capacity to 3              resourceGroup=kargakisosa
INFO[2018-09-10T10:33:10+02:00] waiting for compute-000008 to be ready        resourceGroup=kargakisosa
INFO[2018-09-10T10:34:07+02:00] draining compute-000001                       resourceGroup=kargakisosa
INFO[2018-09-10T10:34:09+02:00] deleting compute-000001                       resourceGroup=kargakisosa
INFO[2018-09-10T10:36:10+02:00] setting ss-compute capacity to 3              resourceGroup=kargakisosa
INFO[2018-09-10T10:40:14+02:00] waiting for compute-000009 to be ready        resourceGroup=kargakisosa
INFO[2018-09-10T10:41:33+02:00] draining compute-000007                       resourceGroup=kargakisosa
INFO[2018-09-10T10:41:36+02:00] deleting compute-000007                       resourceGroup=kargakisosa
INFO[2018-09-10T10:44:38+02:00] starting health check                         resourceGroup=kargakisosa
INFO[2018-09-10T10:44:39+02:00] checking infrastructure service health        resourceGroup=kargakisosa
INFO[2018-09-10T10:44:40+02:00] checking  "prometheus-node-exporter"          resourceGroup=kargakisosa
INFO[2018-09-10T10:44:40+02:00] checking  "sync"                              resourceGroup=kargakisosa
INFO[2018-09-10T10:44:40+02:00] checking  "ovs"                               resourceGroup=kargakisosa
INFO[2018-09-10T10:44:40+02:00] checking  "sdn"                               resourceGroup=kargakisosa
INFO[2018-09-10T10:44:40+02:00] checking  "prometheus"                        resourceGroup=kargakisosa
INFO[2018-09-10T10:44:41+02:00] checking  "docker-registry"                   resourceGroup=kargakisosa
INFO[2018-09-10T10:44:41+02:00] checking  "registry-console"                  resourceGroup=kargakisosa
INFO[2018-09-10T10:44:41+02:00] checking  "router"                            resourceGroup=kargakisosa
INFO[2018-09-10T10:44:41+02:00] checking  "apiserver"                         resourceGroup=kargakisosa
INFO[2018-09-10T10:44:41+02:00] checking  "controller-manager"                resourceGroup=kargakisosa
INFO[2018-09-10T10:44:42+02:00] checking  "asb"                               resourceGroup=kargakisosa
INFO[2018-09-10T10:44:42+02:00] checking  "bootstrap-autoapprover"            resourceGroup=kargakisosa
INFO[2018-09-10T10:44:42+02:00] checking  "apiserver"                         resourceGroup=kargakisosa
INFO[2018-09-10T10:44:42+02:00] checking  "webconsole"                        resourceGroup=kargakisosa
INFO[2018-09-10T10:44:42+02:00] done checking infrastructure service health   resourceGroup=kargakisosa
INFO[2018-09-10T10:44:42+02:00] checking console health                       resourceGroup=kargakisosa
INFO[2018-09-10T10:44:42+02:00] OK                                            resourceGroup=kargakisosa
```
</details>

**Comment by kargakis** [comment]
/hold

once lgtmd, this needs squashing

**Comment by jim-minter** [comment]
lgtm, passed tests, squashing and merging manually because I want to rebase #389 on top.

**Pull request merged**